File: apps/mobile/src/i18n/languages.ts
// Languages the Salarean mobile UI is translated into.
// Source-of-truth list mirrors aoep_shared.languages.SUPPORTED_LANGUAGES so
// content language coverage and UI language coverage stay aligned.
//
// `tier` is "full" when every UI string has a hand-written translation, and
// "fallback" when the app falls back to English for missing keys (so adding
// a language to the picker is non-breaking even before it's fully translated).

export type LocaleCode =
  | "en" | "es" | "fr" | "de" | "it" | "pt" | "nl" | "pl" | "ru" | "uk"
  | "tr" | "ar" | "he" | "hi" | "bn" | "ur" | "fa" | "zh" | "ja" | "ko"
  | "vi" | "th" | "id" | "sw" | "el" | "cs" | "km";

export type LanguageInfo = {
  code: LocaleCode;
  name: string;       // English name
  native: string;     // endonym
  flag: string;
  rtl?: boolean;
  tier: "full" | "fallback";
};

export const LANGUAGES: LanguageInfo[] = [
  { code: "en", name: "English",            native: "English",          flag: "🇬🇧", tier: "full" },
  { code: "es", name: "Spanish",            native: "Español",          flag: "🇪🇸", tier: "full" },
  { code: "fr", name: "French",             native: "Français",         flag: "🇫🇷", tier: "full" },
  { code: "de", name: "German",             native: "Deutsch",          flag: "🇩🇪", tier: "full" },
  { code: "it", name: "Italian",            native: "Italiano",         flag: "🇮🇹", tier: "full" },
  { code: "pt", name: "Portuguese",         native: "Português",        flag: "🇵🇹", tier: "full" },
  { code: "ru", name: "Russian",            native: "Русский",          flag: "🇷🇺", tier: "full" },
  { code: "ar", name: "Arabic",             native: "العربية",            flag: "🇸🇦", rtl: true, tier: "full" },
  { code: "hi", name: "Hindi",              native: "हिन्दी",              flag: "🇮🇳", tier: "full" },
  { code: "zh", name: "Chinese (Mandarin)", native: "中文",              flag: "🇨🇳", tier: "full" },
  { code: "ja", name: "Japanese",           native: "日本語",            flag: "🇯🇵", tier: "full" },
  { code: "ko", name: "Korean",             native: "한국어",             flag: "🇰🇷", tier: "full" },
  { code: "vi", name: "Vietnamese",         native: "Tiếng Việt",       flag: "🇻🇳", tier: "full" },
  { code: "km", name: "Khmer",              native: "ខ្មែរ",            flag: "🇰🇭", tier: "full" },
  // Remaining 13 are wired in the picker but inherit English until a full
  // translation lands. (The strings dictionary cleanly falls through.)
  { code: "nl", name: "Dutch",              native: "Nederlands",       flag: "🇳🇱", tier: "fallback" },
  { code: "pl", name: "Polish",             native: "Polski",           flag: "🇵🇱", tier: "fallback" },
  { code: "uk", name: "Ukrainian",          native: "Українська",       flag: "🇺🇦", tier: "fallback" },
  { code: "tr", name: "Turkish",            native: "Türkçe",           flag: "🇹🇷", tier: "fallback" },
  { code: "he", name: "Hebrew",             native: "עברית",             flag: "🇮🇱", rtl: true, tier: "fallback" },
  { code: "bn", name: "Bengali",            native: "বাংলা",              flag: "🇧🇩", tier: "fallback" },
  { code: "ur", name: "Urdu",               native: "اردو",              flag: "🇵🇰", rtl: true, tier: "fallback" },
  { code: "fa", name: "Persian",            native: "فارسی",            flag: "🇮🇷", rtl: true, tier: "fallback" },
  { code: "th", name: "Thai",               native: "ไทย",               flag: "🇹🇭", tier: "fallback" },
  { code: "id", name: "Indonesian",         native: "Bahasa Indonesia", flag: "🇮🇩", tier: "fallback" },
  { code: "sw", name: "Swahili",            native: "Kiswahili",        flag: "🇰🇪", tier: "fallback" },
  { code: "el", name: "Greek",              native: "Ελληνικά",         flag: "🇬🇷", tier: "fallback" },
  { code: "cs", name: "Czech",              native: "Čeština",          flag: "🇨🇿", tier: "fallback" },
];

export const DEFAULT_LOCALE: LocaleCode = "en";

export function isSupportedLocale(code: string): code is LocaleCode {
  return LANGUAGES.some((l) => l.code === code);
}

export function languageInfo(code: LocaleCode): LanguageInfo {
  return LANGUAGES.find((l) => l.code === code) || LANGUAGES[0];
}

export const RTL_LOCALES: ReadonlySet<LocaleCode> = new Set(
  LANGUAGES.filter((l) => l.rtl).map((l) => l.code)
);

File: subrepos/theodore_audio_translation_lab/src/theodore_audio_translation_lab/languages.py
"""Platform language coverage, importable standalone without aoep_shared."""

from __future__ import annotations

try:
    from aoep_shared.languages import (  # type: ignore
        LANGUAGE_NAMES as _NAMES,
        SUPPORTED_LANGUAGES as _LANGS,
        normalize_language as _normalize,
    )

    SUPPORTED_LANGUAGES: tuple[str, ...] = tuple(_LANGS)
    LANGUAGE_NAMES: dict[str, str] = dict(_NAMES)
    SOURCE = "aoep_shared.languages"

    def normalize_language(value: str, default: str = "") -> str:
        return _normalize(value) or default

except Exception:  # noqa: BLE001 — standalone lab mode
    SUPPORTED_LANGUAGES = (
        "en", "es", "fr", "de", "it", "pt", "nl", "pl", "ru", "uk",
        "tr", "ar", "he", "hi", "bn", "ur", "fa", "zh", "ja", "ko",
        "vi", "th", "id", "sw", "el", "cs", "km",
    )
    LANGUAGE_NAMES = {
        "en": "English", "es": "Spanish", "fr": "French", "de": "German",
        "it": "Italian", "pt": "Portuguese", "nl": "Dutch", "pl": "Polish",
        "ru": "Russian", "uk": "Ukrainian", "tr": "Turkish", "ar": "Arabic",
        "he": "Hebrew", "hi": "Hindi", "bn": "Bengali", "ur": "Urdu",
        "fa": "Persian", "zh": "Chinese (Mandarin)", "ja": "Japanese",
        "ko": "Korean", "vi": "Vietnamese", "th": "Thai",
        "id": "Indonesian", "sw": "Swahili", "el": "Greek", "cs": "Czech",
        "km": "Khmer",
    }
    SOURCE = "embedded"

    def normalize_language(value: str, default: str = "") -> str:
        code = (value or "").strip().lower().split("-")[0]
        return code if code in SUPPORTED_LANGUAGES else default


RTL_LANGUAGES = frozenset({"ar", "he", "ur", "fa"})

# BCP-47 locales for browser/device recognizers.
BCP47: dict[str, str] = {
    "en": "en-US", "es": "es-ES", "fr": "fr-FR", "de": "de-DE",
    "it": "it-IT", "pt": "pt-BR", "nl": "nl-NL", "pl": "pl-PL",
    "ru": "ru-RU", "uk": "uk-UA", "tr": "tr-TR", "ar": "ar-SA",
    "he": "he-IL", "hi": "hi-IN", "bn": "bn-BD", "ur": "ur-PK",
    "fa": "fa-IR", "zh": "zh-CN", "ja": "ja-JP", "ko": "ko-KR",
    "vi": "vi-VN", "th": "th-TH", "id": "id-ID", "sw": "sw-KE",
    "el": "el-GR", "cs": "cs-CZ", "km": "km-KH",
}


def language_rows() -> list[dict]:
    return [
        {
            "code": code,
            "name": LANGUAGE_NAMES[code],
            "bcp47": BCP47[code],
            "rtl": code in RTL_LANGUAGES,
        }
        for code in SUPPORTED_LANGUAGES
    ]

AUTO_LANGUAGE = "auto"


def normalize_input_language(value: str, default: str = "") -> str:
    """Normalize a selected source language, preserving Whisper auto-detect."""
    raw = (value or "").strip().lower()
    if raw in {"auto", "detect", "auto-detect", "automatic"}:
        return AUTO_LANGUAGE
    return normalize_language(value, default=default)


def resolve_detected_language(value: str, default: str = "") -> str:
    """Map Whisper's language code/name output to a platform ISO code."""
    code = normalize_language(value)
    if code:
        return code
    name = (value or "").strip().casefold()
    aliases = {
        "mandarin": "zh",
        "mandarin chinese": "zh",
        "chinese": "zh",
        "farsi": "fa",
        "persian": "fa",
    }
    if name in aliases:
        return aliases[name]
    for lang, display in LANGUAGE_NAMES.items():
        if name == display.casefold() or name == display.split("(", 1)[0].strip().casefold():
            return lang
    return default
